google/api_core/retry/retry_base.py

```diff
@@ -318,26 +309,42 @@ def with_predicate(self, predicate: Callable[[Exception], bool]) -> Self:
         Returns:
             Retry: A new retry instance with the given predicate.
         """
-        return self._replace(predicate=predicate)
+        return type(self)(
+            predicate=predicate,
```

Comment:
Are we allowing `predicate` to be `None` here? If so, we would need to tweak the type annotation, and to add a test for this edge case? (The test might be starting with something with a non-trivial `predicate`, and then verifying that calling `.with_predicate(None)` sets it to `None`?

Reply:
It seems like predicate has always been a required variable

Reply:
Well, in `replace`, we had `predicate: Callable[[Exception], bool] | None = None,`, right?

Reply:
Sorry, I missed this comment

In the _replace helper, `None` represented a variable that isn't replaced (that is, it used the same value as the original Retry object). 

But this  doesn't work if users are passing in `None` for a `with_timeout` and expecting it to disable timeouts, which wasn't supported by the type annotations, but was used in practice. Which was the motivation for this change